libs/editor/egui_editor/src/apple/apple_ffi.rs

```diff
@@ -0,0 +1,117 @@
+        width: screen.physical_width, // TODO get from context or something
+        height: screen.physical_height,
+        present_mode: wgpu::PresentMode::Fifo,
+    };
+    surface.configure(&device, &surface_config);
+    let rpass = egui_wgpu_backend::RenderPass::new(&device, surface_format, 1);
+
+    let context = Context::default();
+    let editor = Editor::default();
+    editor.set_font(&context);
+
+    let mut obj = WgpuEditor {
+        device,
+        queue,
+        surface,
+        adapter,
+        rpass,
+        screen,
+        context,
+        raw_input: Default::default(),
+        editor,
+    };
+
+    obj.frame();
+
+    // TODO we need to free this memory
+    Box::into_raw(Box::new(obj)) as *mut c_void
+}
+
+#[no_mangle]
+pub extern "C" fn draw_editor(obj: *mut c_void) {
+    let obj = unsafe { &mut *(obj as *mut WgpuEditor) };
+    obj.frame();
+}
+
+#[no_mangle]
+pub extern "C" fn resize_editor(obj: *mut c_void, width: f32, height: f32, scale: f32) {
+    let obj = unsafe { &mut *(obj as *mut WgpuEditor) };
+    obj.screen.physical_width = width as u32;
+    obj.screen.physical_height = height as u32;
+    obj.screen.scale_factor = scale;
+}
+
+/// # Safety
+#[no_mangle]
+pub unsafe extern "C" fn key_event(
+    obj: *mut c_void, key_code: u16, shift: bool, ctrl: bool, option: bool, command: bool,
+    pressed: bool, characters: *const c_char,
+) {
+    let obj = &mut *(obj as *mut WgpuEditor);
+
+    let modifiers = egui::Modifiers { alt: option, ctrl, shift, mac_cmd: command, command };
+
+    obj.raw_input.modifiers = modifiers;
+
+    let key = NSKeys::from(key_code).unwrap();
+
+    // Event::Text
+    if pressed && (modifiers.shift_only() || modifiers.is_none()) && key.valid_text() {
+        let text = CStr::from_ptr(characters).to_str().unwrap().to_string();
+        obj.raw_input.events.push(Event::Text(text));
+    }
+
+    // Event::Key
+    let key = key.egui_key().unwrap();
+    obj.raw_input
+        .events
+        .push(Event::Key { key, pressed, modifiers });
+}
+
+async fn request_device(
+    instance: &wgpu::Instance, backend: wgpu::Backends, surface: &wgpu::Surface,
+) -> (wgpu::Adapter, wgpu::Device, wgpu::Queue) {
+    let adapter =
+        wgpu::util::initialize_adapter_from_env_or_default(instance, backend, Some(surface))
+            .await
+            .expect("No suitable GPU adapters found on the system!");
+    let adapter_info = adapter.get_info();
+    println!("Using {} ({:?})", adapter_info.name, adapter_info.backend);
+    let res = adapter
+        .request_device(
+            &wgpu::DeviceDescriptor {
+                label: None,
+                features: adapter.features(),
+                limits: adapter.limits(),
+            },
+            None,
+        )
+        .await;
+    match res {
+        Err(err) => {
+            panic!("request_device failed: {:?}", err);
+        }
+        Ok(tuple) => (adapter, tuple.0, tuple.1),
```

Comment:
Would be clearer to name the tuple's values

Reply:
yeah bit of a nit but would like to see more destructuring assignment from you in general @Parth; aside from getting the `usize` out of our 1-tuple structs like `DocCharOffset`, I view every instance of `.0` or `.1` like this as a code smell

Reply:
yeah this bit was copied and pasted from samples elsewhere, but agreed

Reply:
I'll address shortly